### PR TITLE
Bluetooth: Implement bt_addr_le_is_valid

### PR DESCRIPTION
**Pull request by moonlight83340**
Add a `bool bt_addr_le_is_valid` function that verifies if the address provided is valid.
See https://github.com/zephyrproject-rtos/zephyr/issues/71801#

### CODE REVIEW

File: subsys/bluetooth/host/addr.c [hunk 1]
```diff
@@ -120,3 +120,19 @@ bool bt_addr_le_is_resolved(const bt_addr_le_t *addr)
...
+		return false;
+	}
+
+	if (addr->type != BT_ADDR_LE_PUBLIC && addr->type != BT_ADDR_LE_RANDOM &&
+	    addr->type != BT_ADDR_LE_PUBLIC_ID && addr->type != BT_ADDR_LE_RANDOM_ID) {
+		return false;
+	}
+
+	/* Additional checks for validity can be added here based on the core spec */
+
```
Comment (Thalley) [inline subsys/bluetooth/host/addr.c:136]: IIRC there are some bits of the address that needs to be set/unset in the case of NRPAs and RPA (or something like that). I'll see if I can dig anything specific up

**Review by Thalley — DISMISSED** [review]

**Comment by Thalley** [comment]
> @Thalley Sorry for the ping, I would like to try to develop on the bluetooth area. But I'm new in this part of zephyr and don't really know what should be done. I have done some works on the posix area before. Do you have somes hints ?

So far so good :) 

The next step would be to look for public API functions that take an address as input. 

I would begin by searching all functions in https://github.com/zephyrproject-rtos/zephyr/tree/main/include/zephyr/bluetooth for `bt_addr_le_t *`. Mind that some functions may take the address as a field of another struct like https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/bluetooth/audio/bap.h#L2224-L2260

**Comment by moonlight83340** [comment]
> > @Thalley Sorry for the ping, I would like to try to develop on the bluetooth area. But I'm new in this part of zephyr and don't really know what should be done. I have done some works on the posix area before. Do you have somes hints ?
> 
> So far so good :)
> 
> The next step would be to look for public API functions that take an address as input.
> 
> I would begin by searching all functions in https://github.com/zephyrproject-rtos/zephyr/tree/main/include/zephyr/bluetooth for `bt_addr_le_t *`. Mind that some functions may take the address as a field of another struct like https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/bluetooth/audio/bap.h#L2224-L2260

Thank you, I will take a look ! 😄

File: subsys/bluetooth/host/addr.c [hunk 2]
```diff
@@ -120,3 +120,19 @@ bool bt_addr_le_is_resolved(const bt_addr_le_t *addr)
...
 }
+
+bool bt_addr_le_is_valid(const bt_addr_le_t *addr)
+{
+	if (addr == NULL) {
+		return false;
+	}
+
+	if (addr->type != BT_ADDR_LE_PUBLIC && addr->type != BT_ADDR_LE_RANDOM &&
+	    addr->type != BT_ADDR_LE_PUBLIC_ID && addr->type != BT_ADDR_LE_RANDOM_ID) {
```
Comment (alwa-nordic) [inline subsys/bluetooth/host/addr.c:131]: These are not valid Bluetooth LE address types: `BT_ADDR_LE_PUBLIC_ID` `BT_ADDR_LE_RANDOM_ID`.

Rather, those have a bit set that informs the host that the controller saw a RPA on air, but the controller resolved it for us.

File: subsys/bluetooth/host/addr.c (hunk 1 above)
Comment (alwa-nordic) [reply subsys/bluetooth/host/addr.c:135]: We have a macro for checking random addresses here: https://github.com/zephyrproject-rtos/zephyr/blob/e4747b28a8f3edac3ba8282bb7172898940bcd37/include/zephyr/bluetooth/addr.h#L142-L148

The complete check for random is `type != BT_ADDR_LE_RANDOM || BT_ADDR_IS_RPA(a) || BT_ADDR_IS_NRPA(a) || BT_ADDR_IS_STATIC(a)`.

I think the 'public' namespace is EUI-48. Not sure. But if so, that also has some requirements.

Comment (Thalley) [reply subsys/bluetooth/host/addr.c:135]: > We have a macro for checking random addresses here:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/e4747b28a8f3edac3ba8282bb7172898940bcd37/include/zephyr/bluetooth/addr.h#L142-L148
> 
> The complete check for random is `type != BT_ADDR_LE_RANDOM || BT_ADDR_IS_RPA(a) || BT_ADDR_IS_NRPA(a) || BT_ADDR_IS_STATIC(a)`.
> 
> I think the 'public' namespace is EUI-48. Not sure. But if so, that also has some requirements.

As per the issue (https://github.com/zephyrproject-rtos/zephyr/issues/71801) there are additional requirements for random addresses, such as at least one bit of the random part shall be 0 and one shall be 1. 

For example 40:00:00:00:00:00 and 40:ff:ff:ff:ff:ff are not valid RPAs, but will `BT_ADDR_IS_RPA` will return `true` on those

**Pull request closed without merging**